### ISSUE TITLE
Incorrect const pointers in the kratos core

### ISSUE THREAD

**Issue by EduardGomezEscandell** [KratosMultiphysics/Kratos#9779]
**Description**
There are many methods all over the core with signatures similar to:
```C++
const Type::Pointer T::pGetMember(...) const
```

The presence of a `const` makes it seem as though the pointed object is inmutable, but it is not. Indeed this code would be perfectly valid:
```c++
const T t();
const auto suposedly_const = t.pGetMember();
*suposedly_const += 1;
```
which clearly goes against the spirit of the const getter.

**Solution**
I argue that we should add a new typedef to the pointer definition:
```C++

#define KRATOS_CLASS_POINTER_DEFINITION(a)  \
/* Current */ \
typedef Kratos::shared_ptr<a > Pointer; \
typedef Kratos::shared_ptr<a > SharedPointer; \
typedef Kratos::weak_ptr  <a > WeakPointer; \
typedef Kratos::unique_ptr<a > UniquePointer; \
/* New  */ \
typedef Kratos::shared_ptr<const a > ConstPointer; \
typedef Kratos::shared_ptr<const a > ConstSharedPointer; \
typedef Kratos::weak_ptr  <const a > ConstWeakPointer; \
typedef Kratos::unique_ptr<const a > ConstUniquePointer // Maybe this one can be skipped
```
and something similar with `KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION`

And replace al const pGetters return types to the appropiate const pointer.

**Homework**
I don't mind implementing these changes myself, but I thought I would bring it up before putting in the work.

**Comment by RiccardoRossi** [comment]
pointer to const or const pointer?

https://isocpp.org/wiki/faq/const-correctness#const-ptr-vs-ptr-const

i never completely remember this, but the link is useful

**Comment by EduardGomezEscandell** [comment]
Yes, currently pGetters return const pointers when they should return pointers to const.

**Comment by roigcarlo** [comment]
How would cast work? Is it possible for the compiler to automatically convert `shared<const T>` to `shared<T>`. If not we will break half of the code...

**Comment by EduardGomezEscandell** [comment]
> How would cast work? Is it possible for the compiler to automatically convert shared<const T> to shared<T>. If not we will break half of the code...

Not in this direction (it would drop const qualifier), but yes in the oposite direction:
- `shared<const T>` to `shared<T>`: disallowed
- `shared<T>` to `shared<const T>`: allowed and implicit

**Comment by EduardGomezEscandell** [comment]
Since these changes break API in some contexts, I created https://github.com/KratosMultiphysics/Kratos/pull/9790 to add a deprecation warning and apply the changes when we see fit.
